src/protocols/screencopy.rs

```diff
@@ -0,0 +1,279 @@
+        ] => $crate::protocols::screencopy::ScreencopyManagerState);
+
+        smithay::reexports::wayland_server::delegate_dispatch!($(@< $( $lt $( : $clt $(+ $dlt )* )? ),+ >)? $ty: [
+            smithay::reexports::wayland_protocols_wlr::screencopy::v1::server::zwlr_screencopy_frame_v1::ZwlrScreencopyFrameV1: $crate::protocols::screencopy::ScreencopyFrameState
+        ] => $crate::protocols::screencopy::ScreencopyManagerState);
+    };
+}
+
+#[derive(Clone)]
+pub struct ScreencopyFrameState {
+    pub output: Output,
+    pub rect: Rectangle<i32, Physical>,
+    pub overlay_cursor: bool,
+}
+
+impl<D> Dispatch<ZwlrScreencopyFrameV1, ScreencopyFrameState, D> for ScreencopyManagerState
+where
+    D: Dispatch<ZwlrScreencopyFrameV1, ScreencopyFrameState>,
+    D: ScreencopyHandler,
+    D: 'static,
+{
+    fn request(
+        state: &mut D,
+        _client: &Client,
+        frame: &ZwlrScreencopyFrameV1,
+        request: zwlr_screencopy_frame_v1::Request,
+        data: &ScreencopyFrameState,
+        _display: &DisplayHandle,
+        _data_init: &mut DataInit<'_, D>,
+    ) {
+        let (buffer, send_damage) = match request {
+            zwlr_screencopy_frame_v1::Request::Copy { buffer } => (buffer, false),
+            zwlr_screencopy_frame_v1::Request::CopyWithDamage { buffer } => (buffer, true),
+            zwlr_screencopy_frame_v1::Request::Destroy => return,
+            _ => unreachable!(),
+        };
+
+        state.frame(Screencopy {
+            send_damage,
+            buffer,
+            frame: frame.clone(),
+            frame_state: data.clone(),
+            submitted: false,
+        });
+    }
+}
+
+/// Screencopy frame.
+pub struct Screencopy {
+    frame_state: ScreencopyFrameState,
+    frame: ZwlrScreencopyFrameV1,
+    send_damage: bool,
+    buffer: WlBuffer,
+    submitted: bool,
+}
+
+impl Drop for Screencopy {
+    fn drop(&mut self) {
+        if !self.submitted {
+            self.frame.failed();
+        }
+    }
+}
+
+impl Screencopy {
+    /// Get the target buffer to copy to.
+    pub fn buffer(&self) -> &WlBuffer {
+        &self.buffer
+    }
+
+    /// Get the region which should be copied.
+    pub fn region(&self) -> Rectangle<i32, Physical> {
+        self.frame_state.rect
+    }
+
+    pub fn output(&self) -> &Output {
+        &self.frame_state.output
+    }
+
+    pub fn overlay_cursor(&self) -> bool {
+        self.frame_state.overlay_cursor
+    }
+
+    /// Mark damaged regions of the screencopy buffer.
+    pub fn damage(&mut self, damage: &[Rectangle<i32, Physical>]) {
+        if !self.send_damage {
+            return;
+        }
+
+        for Rectangle { loc, size } in damage {
+            self.frame
+                .damage(loc.x as u32, loc.y as u32, size.w as u32, size.h as u32);
+        }
+    }
+
+    /// Submit the copied content.
+    pub fn submit(mut self) {
```

Comment:
Suggestion to improve the code: since you're passing by value here, you can remove the `self.submitted` field and always send `failed()` on drop, meanwhile in this function just destructure with `let Screencopy { fields } = self;` which doesn't run `drop()`.

Reply:
This doesn't seem to be the case? i.e. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4c08007f797f3559bb36867db18f901f

Potentially, we could do `mem::forget`, which is maybe not desirable? From what i can tell, neither the `ZwlrScreencopyFrameV1` or `WlBuffer` types have a specific `Drop` impl, so it might be just fine, but i think the current way from Catacomb seems more "robust" in not relying on those details

Reply:
Huh, apparently you can't destructure types with Drop?

![image](https://github.com/YaLTeR/niri/assets/1794388/e7b7f0e5-f168-45c6-8065-424085886d8a)

Then just leave as is, yeah.